File: src/lib/taches/tache-recurrence.ts
import { dateInputToUnix, formatCalendarDateFr, unixToDateInput } from "@/lib/dates/calendar-date";

const DAY_SEC = 86400;

export type TacheRecurrenceFreq = "daily" | "weekly" | "monthly" | "yearly";

export interface TacheRecurrence {
  freq: TacheRecurrenceFreq;
  /** Tous les N jours / semaines / mois / ans (défaut 1). */
  interval?: number;
  /** Mensuel / annuel : jour du mois (1–31). */
  day_of_month?: number;
  /** Annuel : mois (1–12). */
  month?: number;
  /** Hebdomadaire : 1=lundi … 7=dimanche. */
  weekdays?: number[];
  /** Fin optionnelle (timestamp Unix minuit UTC). */
  until?: number | null;
}

export const TACHE_RECURRENCE_FREQ_OPTIONS: {
  value: TacheRecurrenceFreq;
  label: string;
  intervalUnit: string;
}[] = [
  { value: "daily", label: "Quotidienne", intervalUnit: "jour(s)" },
  { value: "weekly", label: "Hebdomadaire", intervalUnit: "semaine(s)" },
  { value: "monthly", label: "Mensuelle", intervalUnit: "mois" },
  { value: "yearly", label: "Annuelle", intervalUnit: "an(s)" },
];

export const TACHE_WEEKDAY_OPTIONS: { value: number; label: string }[] = [
  { value: 1, label: "Lun" },
  { value: 2, label: "Mar" },
  { value: 3, label: "Mer" },
  { value: 4, label: "Jeu" },
  { value: 5, label: "Ven" },
  { value: 6, label: "Sam" },
  { value: 7, label: "Dim" },
];

export const TACHE_MONTH_OPTIONS: { value: number; label: string }[] = [
  { value: 1, label: "Janvier" },
  { value: 2, label: "Février" },
  { value: 3, label: "Mars" },
  { value: 4, label: "Avril" },
  { value: 5, label: "Mai" },
  { value: 6, label: "Juin" },
  { value: 7, label: "Juillet" },
  { value: 8, label: "Août" },
  { value: 9, label: "Septembre" },
  { value: 10, label: "Octobre" },
  { value: 11, label: "Novembre" },
  { value: 12, label: "Décembre" },
];

export function isActiveRecurrence(
  rec: TacheRecurrence | null | undefined
): rec is TacheRecurrence {
  return rec != null && rec.freq != null && rec.freq.length > 0;
}

function utcNaiveDate(ts: number): { y: number; m: number; d: number } {
  const date = new Date(ts * 1000);
  return {
    y: date.getUTCFullYear(),
    m: date.getUTCMonth() + 1,
    d: date.getUTCDate(),
  };
}

function unixFromUtc(y: number, m: number, d: number): number {
  return Math.floor(Date.UTC(y, m - 1, d) / 1000);
}

function isoWeekday(ts: number): number {
  const date = new Date(ts * 1000);
  const js = date.getUTCDay(); // 0=dim
  return js === 0 ? 7 : js;
}

function daysInMonth(year: number, month: number): number {
  return new Date(Date.UTC(year, month, 0)).getUTCDate();
}

function clampDayInMonth(year: number, month: number, day: number): number {
  return Math.min(Math.max(day, 1), daysInMonth(year, month));
}

function addMonths(year: number, month: number, delta: number): { y: number; m: number } {
  const total = year * 12 + (month - 1) + delta;
  const y = Math.floor(total / 12);
  const m = (total % 12) + 1;
  return { y, m };
}

function nextDaily(fromTs: number, interval: number): number {
  return fromTs + interval * DAY_SEC;
}

function nextWeekly(
  fromTs: number,
  interval: number,
  weekdays?: number[]
): number | null {
  const step = Math.max(interval, 1);
  if (!weekdays?.length) {
    return fromTs + 7 * step * DAY_SEC;
  }
  if (weekdays.length === 1) {
    const target = Math.min(Math.max(weekdays[0], 1), 7);
    if (isoWeekday(fromTs) === target) {
      return fromTs + 7 * step * DAY_SEC;
    }
    for (let offset = 1; offset <= 370; offset++) {
      const candidate = fromTs + offset * DAY_SEC;
      if (isoWeekday(candidate) === target) return candidate;
    }
    return null;
  }
  const normalized = weekdays.map((d) => Math.min(Math.max(d, 1), 7));
  for (let offset = 1; offset <= 370; offset++) {
    const candidate = fromTs + offset * DAY_SEC;
    if (normalized.includes(isoWeekday(candidate))) return candidate;
  }
  return null;
}

function nextMonthly(fromTs: number, interval: number, dayOfMonth: number): number {
  const { y, m, d } = utcNaiveDate(fromTs);
  const { y: y2, m: m2 } = addMonths(y, m, Math.max(interval, 1));
  const day =
    dayOfMonth > 0 ? clampDayInMonth(y2, m2, dayOfMonth) : d;
  return unixFromUtc(y2, m2, day);
}

function nextYearly(
  fromTs: number,
  interval: number,
  month: number,
  dayOfMonth: number
): number {
  const { y, d } = utcNaiveDate(fromTs);
  const y2 = y + Math.max(interval, 1);
  const mo = Math.min(Math.max(month, 1), 12);
  const day = dayOfMonth > 0 ? clampDayInMonth(y2, mo, dayOfMonth) : d;
  return unixFromUtc(y2, mo, day);
}

/** Prochaine échéance après l'ancre (aligné backend Rust). */
export function nextTacheOccurrence(
  fromTs: number,
  rec: TacheRecurrence
): number | null {
  if (!isActiveRecurrence(rec)) return null;
  const interval = Math.max(rec.interval ?? 1, 1);
  let next: number | null = null;
  switch (rec.freq) {
    case "daily":
      next = nextDaily(fromTs, interval);
      break;
    case "weekly":
      next = nextWeekly(fromTs, interval, rec.weekdays);
      break;
    case "monthly": {
      const anchor = utcNaiveDate(fromTs);
      const dom = rec.day_of_month ?? anchor.d;
      next = nextMonthly(fromTs, interval, dom);
      break;
    }
    case "yearly": {
      const anchor = utcNaiveDate(fromTs);
      const mo = rec.month ?? anchor.m;
      const dom = rec.day_of_month ?? anchor.d;
      next = nextYearly(fromTs, interval, mo, dom);
      break;
    }
    default:
      return null;
  }
  if (next == null) return null;
  if (rec.until != null && next > rec.until) return null;
  return next;
}

/** Règle par défaut à partir de la date d'échéance saisie. */
export function defaultRecurrenceFromEcheance(
  dateEcheance: string
): TacheRecurrence {
  const ts = dateInputToUnix(dateEcheance);
  const anchor = ts != null ? utcNaiveDate(ts) : utcNaiveDate(Math.floor(Date.now() / 1000));
  return {
    freq: "monthly",
    interval: 1,
    day_of_month: anchor.d,
  };
}

export function formatRecurrenceLabel(rec: TacheRecurrence): string {
  const interval = Math.max(rec.interval ?? 1, 1);
  const every =
    interval === 1 ? "" : `Tous les ${interval} `;
  switch (rec.freq) {
    case "daily":
      return interval === 1 ? "Quotidienne" : `${every}jours`;
    case "weekly": {
      const days =
        rec.weekdays?.length
          ? rec.weekdays
              .map((d) => TACHE_WEEKDAY_OPTIONS.find((o) => o.value === d)?.label ?? "?")
              .join(", ")
          : "semaine";
      return interval === 1
        ? `Hebdo · ${days}`
        : `${every}sem. · ${days}`;
    }
    case "monthly": {
      const dom = rec.day_of_month ?? 1;
      return interval === 1
        ? `Mensuelle · le ${dom}`
        : `${every}mois · le ${dom}`;
    }
    case "yearly": {
      const mo =
        TACHE_MONTH_OPTIONS.find((o) => o.value === rec.month)?.label ?? "?";
      const dom = rec.day_of_month ?? 1;
      return interval === 1
        ? `Annuelle · ${dom} ${mo}`
        : `${every}ans · ${dom} ${mo}`;
    }
    default:
      return "Récurrente";
  }
}

export function formatNextOccurrencePreview(
  dateEcheance: string,
  rec: TacheRecurrence | null
): string | null {
  if (!rec || !isActiveRecurrence(rec)) return null;
  const anchor = dateInputToUnix(dateEcheance);
  if (anchor == null) return null;
  const next = nextTacheOccurrence(anchor, rec);
  if (next == null) return "Fin de récurrence atteinte";
  return `Après validation : ${formatCalendarDateFr(next)}`;
}

export function isoWeekdayFromDateInput(dateInput: string): number {
  const ts = dateInputToUnix(dateInput);
  if (ts == null) return 1;
  return isoWeekday(ts);
}

export function dayOfMonthFromDateInput(dateInput: string): number | null {
  const ts = dateInputToUnix(dateInput);
  if (ts == null) return null;
  return utcNaiveDate(ts).d;
}

export interface RecurrenceEcheanceMismatch {
  message: string;
  alignedDate: string;
  alignedLabel: string;
}

/** Première échéance incohérente avec la règle (ex. le 23 vs « le 1 du mois »). */
export function detectRecurrenceEcheanceMismatch(
  dateEcheance: string,
  rec: TacheRecurrence | null,
  recurrenceEnabled: boolean
): RecurrenceEcheanceMismatch | null {
  if (!recurrenceEnabled || !rec || !isActiveRecurrence(rec)) return null;
  const ts = dateInputToUnix(dateEcheance);
  if (ts == null) return null;

  const alignedDate = alignEcheanceToRecurrence(dateEcheance, rec);
  if (alignedDate === dateEcheance) return null;

  const alignedLabel = formatCalendarDateFr(dateInputToUnix(alignedDate)!);

  switch (rec.freq) {
    case "monthly": {
      const dom = rec.day_of_month ?? dayOfMonthFromDateInput(dateEcheance) ?? 1;
      const currentDom = dayOfMonthFromDateInput(dateEcheance);
      return {
        message: `La première échéance est le ${currentDom}, alors que la série est calée sur le ${dom} de chaque mois.`,
        alignedDate,
        alignedLabel,
      };
    }
    case "yearly": {
      const mo =
        TACHE_MONTH_OPTIONS.find((o) => o.value === rec.month)?.label ?? "?";
      const dom = rec.day_of_month ?? 1;
      return {
        message: `La première échéance ne correspond pas au ${dom} ${mo} de la série annuelle.`,
        alignedDate,
        alignedLabel,
      };
    }
    case "weekly": {
      if (!rec.weekdays?.length) return null;
      const labels = rec.weekdays
        .map((d) => TACHE_WEEKDAY_OPTIONS.find((o) => o.value === d)?.label ?? "?")
        .join(", ");
      return {
        message: `La première échéance n'est pas un ${labels}.`,
        alignedDate,
        alignedLabel,
      };
    }
    default:
      return null;
  }
}

/** Prochaine date calée sur la règle, à partir de la première échéance saisie. */
export function alignEcheanceToRecurrence(
  dateEcheance: string,
  rec: TacheRecurrence
): string {
  const fromTs = dateInputToUnix(dateEcheance);
  if (fromTs == null) return dateEcheance;

  switch (rec.freq) {
    case "monthly": {
      const dom = rec.day_of_month ?? utcNaiveDate(fromTs).d;
      return nextMonthlyDomOnOrAfter(dateEcheance, dom);
    }
    case "yearly": {
      const month = rec.month ?? utcNaiveDate(fromTs).m;
      const dom = rec.day_of_month ?? utcNaiveDate(fromTs).d;
      return nextYearlyOnOrAfter(dateEcheance, month, dom);
    }
    case "weekly": {
      const weekdays = rec.weekdays?.length
        ? rec.weekdays
        : [isoWeekday(fromTs)];
      return nextWeekdayOnOrAfter(dateEcheance, weekdays);
    }
    default:
      return dateEcheance;
  }
}

function monthlyDomTs(year: number, month: number, dayOfMonth: number): number {
  const d = clampDayInMonth(year, month, dayOfMonth);
  return unixFromUtc(year, month, d);
}

function nextMonthlyDomOnOrAfter(dateInput: string, dayOfMonth: number): string {
  const fromTs = dateInputToUnix(dateInput);
  if (fromTs == null) return dateInput;
  const anchor = utcNaiveDate(fromTs);
  let candidate = monthlyDomTs(anchor.y, anchor.m, dayOfMonth);
  if (candidate >= fromTs) return unixToDateInput(candidate);
  const { y, m } = addMonths(anchor.y, anchor.m, 1);
  return unixToDateInput(monthlyDomTs(y, m, dayOfMonth));
}

function nextYearlyOnOrAfter(
  dateInput: string,
  month: number,
  dayOfMonth: number
): string {
  const fromTs = dateInputToUnix(dateInput);
  if (fromTs == null) return dateInput;
  const anchor = utcNaiveDate(fromTs);
  const mo = Math.min(Math.max(month, 1), 12);
  let candidate = monthlyDomTs(anchor.y, mo, dayOfMonth);
  if (candidate >= fromTs) return unixToDateInput(candidate);
  return unixToDateInput(monthlyDomTs(anchor.y + 1, mo, dayOfMonth));
}

function nextWeekdayOnOrAfter(dateInput: string, weekdays: number[]): string {
  const fromTs = dateInputToUnix(dateInput);
  if (fromTs == null) return dateInput;
  const normalized = [...new Set(weekdays.map((d) => Math.min(Math.max(d, 1), 7)))].sort(
    (a, b) => a - b
  );
  if (normalized.includes(isoWeekday(fromTs))) return dateInput;
  for (let offset = 1; offset <= 370; offset++) {
    const candidate = fromTs + offset * DAY_SEC;
    if (normalized.includes(isoWeekday(candidate))) {
      return unixToDateInput(candidate);
    }
  }
  return dateInput;
}

File: src-tauri/src/database/tache_recurrence.rs
//! Calcul des occurrences suivantes pour les tâches récurrentes (UTC calendaire).

use chrono::{Datelike, NaiveDate};
use serde::{Deserialize, Serialize};

const DAY_SEC: i64 = 86400;

/// Règle de récurrence sérialisée en JSON (`taches.recurrence`).
#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq, Eq)]
#[serde(default)]
pub struct TacheRecurrence {
    /// `daily` | `weekly` | `monthly` | `yearly`
    pub freq: String,
    /// Tous les N jours / semaines / mois / ans (défaut 1).
    pub interval: Option<u32>,
    /// Mensuel / annuel : jour du mois (1–31, réduit au dernier jour si besoin).
    pub day_of_month: Option<u32>,
    /// Annuel : mois (1–12).
    pub month: Option<u32>,
    /// Hebdomadaire : jours ISO 1=lundi … 7=dimanche.
    pub weekdays: Option<Vec<u32>>,
    /// Fin optionnelle (timestamp Unix, minuit UTC inclus).
    pub until: Option<i64>,
}

impl TacheRecurrence {
    pub fn is_active(&self) -> bool {
        !self.freq.is_empty() && self.freq != "none"
    }
}

pub fn parse_recurrence_json(raw: Option<&str>) -> Option<TacheRecurrence> {
    let s = raw?.trim();
    if s.is_empty() {
        return None;
    }
    let rec: TacheRecurrence = serde_json::from_str(s).ok()?;
    if rec.is_active() {
        Some(rec)
    } else {
        None
    }
}

pub fn serialize_recurrence_json(rec: Option<&TacheRecurrence>) -> Option<String> {
    let rec = rec?;
    if !rec.is_active() {
        return None;
    }
    serde_json::to_string(rec).ok()
}

fn utc_naive_date(ts: i64) -> NaiveDate {
    let days = (ts.max(0) / 86400) as i32;
    NaiveDate::from_num_days_from_ce_opt(719163 + days)
        .unwrap_or_else(|| NaiveDate::from_ymd_opt(1970, 1, 1).unwrap())
}

fn unix_from_naive(d: NaiveDate) -> i64 {
    let days = d.num_days_from_ce() - 719163;
    days as i64 * DAY_SEC
}

fn iso_weekday(d: NaiveDate) -> u32 {
    d.weekday().num_days_from_monday() + 1
}

fn days_in_month(year: i32, month: u32) -> u32 {
    NaiveDate::from_ymd_opt(
        year,
        if month >= 12 { 12 } else { month + 1 },
        1,
    )
    .and_then(|d| d.pred_opt())
    .map(|d| d.day())
    .unwrap_or(28)
}

fn clamp_day_in_month(year: i32, month: u32, day: u32) -> u32 {
    let max = days_in_month(year, month);
    day.clamp(1, max)
}

fn add_months(year: i32, month: u32, delta: u32) -> (i32, u32) {
    let total = (year as i64) * 12 + (month as i64 - 1) + delta as i64;
    let y = (total / 12) as i32;
    let m = (total % 12 + 12) % 12 + 1;
    (y, m as u32)
}

fn next_daily(from_ts: i64, interval: u32) -> i64 {
    from_ts + interval as i64 * DAY_SEC
}

fn next_weekly(from_ts: i64, interval: u32, weekdays: Option<&[u32]>) -> Option<i64> {
    let interval = interval.max(1);
    match weekdays {
        None | Some([]) => Some(from_ts + 7 * interval as i64 * DAY_SEC),
        Some(days) if days.len() == 1 => {
            let target = days[0].clamp(1, 7);
            let from_date = utc_naive_date(from_ts);
            if iso_weekday(from_date) == target {
                return Some(from_ts + 7 * interval as i64 * DAY_SEC);
            }
            let mut d = from_date.succ_opt()?;
            for _ in 0..370 {
                if iso_weekday(d) == target {
                    return Some(unix_from_naive(d));
                }
                d = d.succ_opt()?;
            }
            None
        }
        Some(days) => {
            let normalized: Vec<u32> = days.iter().copied().map(|d| d.clamp(1, 7)).collect();
            let from_date = utc_naive_date(from_ts);
            let mut d = from_date.succ_opt()?;
            for _ in 0..370 {
                if normalized.contains(&iso_weekday(d)) {
                    return Some(unix_from_naive(d));
                }
                d = d.succ_opt()?;
            }
            None
        }
    }
}

fn next_monthly(from_ts: i64, interval: u32, day_of_month: u32) -> Option<i64> {
    let from_date = utc_naive_date(from_ts);
    let (y, m) = (from_date.year(), from_date.month());
    let (y2, m2) = add_months(y, m, interval.max(1));
    let day = if day_of_month == 0 {
        from_date.day()
    } else {
        clamp_day_in_month(y2, m2, day_of_month)
    };
    Some(unix_from_naive(NaiveDate::from_ymd_opt(y2, m2, day)?))
}

fn next_yearly(from_ts: i64, interval: u32, month: u32, day_of_month: u32) -> Option<i64> {
    let from_date = utc_naive_date(from_ts);
    let y = from_date.year() + interval.max(1) as i32;
    let m = month.clamp(1, 12);
    let day = if day_of_month == 0 {
        from_date.day()
    } else {
        clamp_day_in_month(y, m, day_of_month)
    };
    Some(unix_from_naive(NaiveDate::from_ymd_opt(y, m, day)?))
}

/// Prochaine échéance strictement après l'ancre (échéance de la tâche terminée).
pub fn next_occurrence(from_ts: i64, rec: &TacheRecurrence) -> Option<i64> {
    if !rec.is_active() {
        return None;
    }
    let interval = rec.interval.unwrap_or(1).max(1);
    let next = match rec.freq.as_str() {
        "daily" => Some(next_daily(from_ts, interval)),
        "weekly" => next_weekly(from_ts, interval, rec.weekdays.as_deref()),
        "monthly" => {
            let dom = rec.day_of_month.unwrap_or(utc_naive_date(from_ts).day());
            next_monthly(from_ts, interval, dom)
        }
        "yearly" => {
            let month = rec.month.unwrap_or(utc_naive_date(from_ts).month());
            let dom = rec
                .day_of_month
                .unwrap_or(utc_naive_date(from_ts).day());
            next_yearly(from_ts, interval, month, dom)
        }
        _ => None,
    }?;
    if let Some(until) = rec.until {
        if next > until {
            return None;
        }
    }
    Some(next)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn ts(y: i32, m: u32, d: u32) -> i64 {
        unix_from_naive(NaiveDate::from_ymd_opt(y, m, d).unwrap())
    }

    #[test]
    fn monthly_day_two_advances_to_next_month() {
        let rec = TacheRecurrence {
            freq: "monthly".into(),
            interval: Some(1),
            day_of_month: Some(2),
            ..Default::default()
        };
        assert_eq!(next_occurrence(ts(2026, 3, 2), &rec), Some(ts(2026, 4, 2)));
        assert_eq!(next_occurrence(ts(2026, 1, 2), &rec), Some(ts(2026, 2, 2)));
    }

    #[test]
    fn monthly_day_31_clamps_in_february() {
        let rec = TacheRecurrence {
            freq: "monthly".into(),
            interval: Some(1),
            day_of_month: Some(31),
            ..Default::default()
        };
        assert_eq!(next_occurrence(ts(2026, 1, 31), &rec), Some(ts(2026, 2, 28)));
    }

    #[test]
    fn weekly_single_weekday_every_two_weeks() {
        let rec = TacheRecurrence {
            freq: "weekly".into(),
            interval: Some(2),
            weekdays: Some(vec![1]), // lundi
            ..Default::default()
        };
        // 2026-06-01 is Monday
        assert_eq!(next_occurrence(ts(2026, 6, 1), &rec), Some(ts(2026, 6, 15)));
    }

    #[test]
    fn weekly_multiple_weekdays_picks_next_day() {
        let rec = TacheRecurrence {
            freq: "weekly".into(),
            interval: Some(1),
            weekdays: Some(vec![1, 5]), // lun, ven
            ..Default::default()
        };
        // 2026-06-01 lun → ven 5
        assert_eq!(next_occurrence(ts(2026, 6, 1), &rec), Some(ts(2026, 6, 5)));
        // ven 5 → lun 8
        assert_eq!(next_occurrence(ts(2026, 6, 5), &rec), Some(ts(2026, 6, 8)));
    }

    #[test]
    fn daily_interval_respected() {
        let rec = TacheRecurrence {
            freq: "daily".into(),
            interval: Some(3),
            ..Default::default()
        };
        assert_eq!(next_occurrence(ts(2026, 6, 1), &rec), Some(ts(2026, 6, 4)));
    }

    #[test]
    fn until_stops_recurrence() {
        let rec = TacheRecurrence {
            freq: "monthly".into(),
            interval: Some(1),
            day_of_month: Some(1),
            until: Some(ts(2026, 3, 1)),
            ..Default::default()
        };
        assert_eq!(next_occurrence(ts(2026, 1, 1), &rec), Some(ts(2026, 2, 1)));
        assert_eq!(next_occurrence(ts(2026, 2, 1), &rec), Some(ts(2026, 3, 1)));
        assert_eq!(next_occurrence(ts(2026, 3, 1), &rec), None);
    }
}
